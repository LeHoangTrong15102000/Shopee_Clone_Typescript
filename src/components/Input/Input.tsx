// Khi mà dùng function thì hả nên import React
import { InputHTMLAttributes, useState } from 'react'
import type { UseFormRegister, RegisterOptions, FieldValues, FieldPath } from 'react-hook-form'

// Mặc định là chúng ta sẽ lấy là FieldValues để nó không có báo lỗi nữa
// TFieldValue nó kế thừa từ FieldValues mặc định chúng ta sẽ lấy TFieldValues, giá trị mặc định có thể gán hoặc không gán đều không sao cả

// Thằng TFieldValues bên trong interface này thì nó chỉ được dùng trong phạm vi của interface thôi
// interface Props<TFieldValues extends FieldValues = FieldValues> extends InputHTMLAttributes<HTMLInputElement> {
//   errorMessage?: string
//   // name: string // để truyền vào tham số thứ nhất cho register
//   classNameInput?: string
//   classNameError?: string
//   classNameEye?: string
//   register?: UseFormRegister<TFieldValues> // có hay không cũng được
//   rules?: RegisterOptions // Là những options trong tham số thứ 2 của register{...}
//   // autoComplete?: string
//   name: FieldPath<TFieldValues>
// }

interface Props<
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
> extends InputHTMLAttributes<HTMLInputElement> {
  errorMessage?: string
  // name: string // để truyền vào tham số thứ nhất cho register
  classNameInput?: string
  classNameError?: string
  classNameEye?: string
  register?: UseFormRegister<TFieldValues> // có hay không cũng được
  rules?: RegisterOptions // Là những options trong tham số thứ 2 của register{...}
  // autoComplete?: string
  name: TName
}

// Cái Generic Type nó vẫn đóng vai trò là cầu nối dữ liệu giũa `nane` và `register`

// arrow function thì phải viết Generic type như thế này, muốn chặt chẽ hơn thì cho thêm = FieldValues vào
const Input = <
  TFieldValues extends FieldValues = FieldValues,
  TName extends FieldPath<TFieldValues> = FieldPath<TFieldValues>
>({
  errorMessage,
  className,
  name,
  register,
  rules,
  classNameInput = 'w-full rounded-sm border border-gray-300 p-3 shadow-sm outline-none focus:border-gray-500',
  classNameError = 'mt-1 min-h-[1.25rem] text-sm text-red-600',
  classNameEye = 'absolute right-[8px] top-[9px] h-6 w-6 cursor-pointer',
  ...rest
}: // TFieldValues ở đây truyền thông qua
Props<TFieldValues, TName>) => {
  // state để lưu trữ việc hiển thị con mắt
  const [openEye, setOpenEye] = useState(false)
  const registerResult = register && name ? register(name, rules as any) : null // {} // làm như này để tái sử dụng component Input ở các nơi khác nhau

  // func toggle eye
  const handleToggleEye = () => {
    setOpenEye((prev) => !prev)
  }

  // func handle việc hiển thị giá trị ô Input khi mà toggleEye
  const handleTypeToggleEye = () => {
    if (rest.type === 'password') {
      return openEye ? 'text' : 'password'
    }
    // Còn không phải là `password` thì return lại rest.type
    return rest.type
  }

  // Tính toán className với border đỏ khi có error
  const getInputClassName = () => {
    if (errorMessage) {
      return classNameInput.replace('border-gray-300', 'border-red-600')
    }
    return classNameInput
  }

  // Tạo unique ID cho error message
  const errorId = errorMessage ? `${name}-error` : undefined

  return (
    <div className={className}>
      {rest.value && (
        <label
          className='absolute top-[-12px] left-[8px] block animate-slide-top-sm bg-white px-1 py-1 text-[12px] italic text-gray-700 transition-all duration-300 ease-out'
          htmlFor={name}
        >
          {name?.slice(0, 1).toUpperCase() + name?.slice(1)}
        </label>
      )}
      <input
        className={getInputClassName()}
        // Tự sinh ra cho chúng ta với name là email, nó sẽ tự sinh ra cho chúng ta nên không cần truyền vào
        {...registerResult}
        {...rest}
        // Ensure name and type are passed through properly
        name={name}
        type={handleTypeToggleEye()}
        // ARIA attributes for accessibility
        aria-invalid={errorMessage ? 'true' : 'false'}
        aria-describedby={errorId}
      />
      {/* Eye-open */}
      {rest.type === 'password' && openEye && (
        <svg
          xmlns='http://www.w3.org/2000/svg'
          fill='none'
          viewBox='0 0 24 24'
          strokeWidth={1.5}
          stroke='currentColor'
          className={classNameEye}
          onClick={handleToggleEye}
        >
          <path
            strokeLinecap='round'
            strokeLinejoin='round'
            d='M2.036 12.322a1.012 1.012 0 010-.639C3.423 7.51 7.36 4.5 12 4.5c4.638 0 8.573 3.007 9.963 7.178.07.207.07.431 0 .639C20.577 16.49 16.64 19.5 12 19.5c-4.638 0-8.573-3.007-9.963-7.178z'
          />
          <path strokeLinecap='round' strokeLinejoin='round' d='M15 12a3 3 0 11-6 0 3 3 0 016 0z' />
        </svg>
      )}
      {rest.type === 'password' && !openEye && (
        <svg
          xmlns='http://www.w3.org/2000/svg'
          fill='none'
          viewBox='0 0 24 24'
          strokeWidth={1.5}
          stroke='currentColor'
          className={classNameEye}
          onClick={handleToggleEye}
        >
          <path
            strokeLinecap='round'
            strokeLinejoin='round'
            d='M3.98 8.223A10.477 10.477 0 001.934 12C3.226 16.338 7.244 19.5 12 19.5c.993 0 1.953-.138 2.863-.395M6.228 6.228A10.45 10.45 0 0112 4.5c4.756 0 8.773 3.162 10.065 7.498a10.523 10.523 0 01-4.293 5.774M6.228 6.228L3 3m3.228 3.228l3.65 3.65m7.894 7.894L21 21m-3.228-3.228l-3.65-3.65m0 0a3 3 0 10-4.243-4.243m4.242 4.242L9.88 9.88'
          />
        </svg>
      )}
      {/* Eye-close */}
      {/* cho m-height để khi mà không có lỗi thì nó vẫn chiếm được vị trí ở đó */}
      <div className={classNameError} id={errorId}>
        {errorMessage}
      </div>
    </div>
  )
}

export default Input
