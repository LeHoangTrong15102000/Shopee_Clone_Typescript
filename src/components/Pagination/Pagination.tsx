// Sử dụng thuật toán range 2 để tạo ra pagination cho app

import classNames from 'classnames'
import { Link, createSearchParams } from 'react-router-dom'
import path from 'src/constant/path'
import { QueryConfig } from 'src/hooks/useQueryConfig'

interface Props {
  queryConfig?: QueryConfig
  pageSize?: number
  basePath?: string
}

const RANGE = 2
const Pagination = ({ queryConfig = {}, pageSize = 20, basePath = path.home }: Props) => {
  // Handle missing or invalid queryConfig
  const safeQueryConfig = queryConfig || {}
  const page = Number(safeQueryConfig.page) || 1
  const safePage = !isNaN(page) && page > 0 ? page : 1
  const safePageSize = pageSize && pageSize > 0 ? pageSize : 20

  // Viết function handle việc paginate cho trang sẽ được thực hiện ở đây
  const renderPagination = () => {
    let dotAfter = false
    let dotBefore = false
    const renderDotBefore = (index: number) => {
      if (!dotBefore) {
        dotBefore = true
        return (
          <span className='flex items-center justify-center border bg-white px-3 py-2 shadow-sm' key={index}>
            ...
          </span>
        )
      }
      return null
    }
    const renderDotAfter = (index: number) => {
      if (!dotAfter) {
        dotAfter = true
        return (
          <span className='flex items-center justify-center border bg-white px-3 py-2 shadow-sm' key={index}>
            ...
          </span>
        )
      }
      return null
    }
    return Array(safePageSize)
      .fill(0)
      .map((_, index) => {
        const pageNumber = index + 1
        // Trường hợp đầu tiênm, điều kiện để return về ...
        if (safePage <= RANGE * 2 + 1 && pageNumber > safePage + RANGE && pageNumber < safePageSize - RANGE + 1) {
          return renderDotAfter(index)
        } else if (safePage > RANGE * 2 + 1 && safePage < safePageSize - RANGE * 2) {
          if (pageNumber < safePage - RANGE && pageNumber > RANGE) {
            return renderDotBefore(index)
          } else if (pageNumber > safePage + RANGE && pageNumber < safePageSize - RANGE + 1) {
            return renderDotAfter(index)
          }
        } else if (safePage >= safePageSize - RANGE * 2 && pageNumber > RANGE && pageNumber < safePage - RANGE) {
          return renderDotBefore(index)
        }
        // Nó sẽ trả về một cái mảng chứa các JSX là các thẻ Link
        return (
          <Link
            to={{
              pathname: basePath,
              search: createSearchParams({
                ...safeQueryConfig,
                page: pageNumber.toString() // cần type string
              }).toString()
            }}
            className={classNames('flex cursor-pointer items-center justify-center px-4 py-2 text-[18px]', {
              'bg-[#ee4d2d] text-white hover:bg-[#ee4d2d]': pageNumber === safePage,
              'border-transparent text-black/50 hover:text-[#ee4e2d]': pageNumber !== safePage
            })}
            key={index}
          >
            {pageNumber}
          </Link>
        )
      })
  }
  return (
    <nav
      className='mt-6 flex flex-wrap items-center justify-center'
      role='navigation'
      aria-label='Pagination Navigation'
    >
      {safePage === 1 ? (
        <span className='flex cursor-not-allowed items-center justify-center rounded-tl-sm rounded-bl-sm border-transparent px-4 py-2 '>
          <svg
            xmlns='http://www.w3.org/2000/svg'
            fill='none'
            viewBox='0 0 24 24'
            strokeWidth={1.5}
            stroke='currentColor'
            className='h-5 w-5 text-[rgba(0,0,0,0.6)]'
          >
            <path strokeLinecap='round' strokeLinejoin='round' d='M15.75 19.5L8.25 12l7.5-7.5' />
          </svg>
        </span>
      ) : (
        <Link
          to={{
            pathname: basePath,
            search: createSearchParams({
              ...safeQueryConfig,
              page: (safePage - 1).toString()
            }).toString()
          }}
          className='flex cursor-pointer items-center justify-center rounded-tl-sm rounded-bl-sm border-transparent px-4 py-2 '
          aria-label='Go to previous page'
        >
          <svg
            xmlns='http://www.w3.org/2000/svg'
            fill='none'
            viewBox='0 0 24 24'
            strokeWidth={1.5}
            stroke='currentColor'
            className='h-5 w-5 text-[rgba(0,0,0,0.6)]'
          >
            <path strokeLinecap='round' strokeLinejoin='round' d='M15.75 19.5L8.25 12l7.5-7.5' />
          </svg>
        </Link>
      )}
      {/* {page === } */}
      {renderPagination()}
      {safePage === safePageSize ? (
        <span className='flex cursor-not-allowed items-center justify-center rounded-tr-sm rounded-br-sm border-transparent  px-4 py-2'>
          <svg
            xmlns='http://www.w3.org/2000/svg'
            fill='none'
            viewBox='0 0 24 24'
            strokeWidth={1.5}
            stroke='currentColor'
            className='h-5 w-5 text-[rgba(0,0,0,.6)] '
          >
            <path strokeLinecap='round' strokeLinejoin='round' d='M8.25 4.5l7.5 7.5-7.5 7.5' />
          </svg>
        </span>
      ) : (
        <Link
          to={{
            pathname: basePath,
            search: createSearchParams({
              ...safeQueryConfig,
              page: (safePage + 1).toString()
            }).toString()
          }}
          className='flex cursor-pointer items-center justify-center rounded-tr-sm rounded-br-sm border-transparent  px-4 py-2 '
          aria-label='Go to next page'
        >
          <svg
            xmlns='http://www.w3.org/2000/svg'
            fill='none'
            viewBox='0 0 24 24'
            strokeWidth={1.5}
            stroke='currentColor'
            className='h-5 w-5 text-[rgba(0,0,0,.6)] '
          >
            <path strokeLinecap='round' strokeLinejoin='round' d='M8.25 4.5l7.5 7.5-7.5 7.5' />
          </svg>
        </Link>
      )}
    </nav>
  )
}

export default Pagination
