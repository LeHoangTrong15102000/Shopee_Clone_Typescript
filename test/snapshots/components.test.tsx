import { describe, test, expect } from 'vitest'
import { render } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import React from 'react'
import Input from '../../src/components/Input'
import Button from '../../src/components/Button'
import Pagination from '../../src/components/Pagination'

// Wrapper component để cung cấp context cần thiết
const TestWrapper = ({ children }: { children: React.ReactNode }) => <BrowserRouter>{children}</BrowserRouter>

describe('Component Snapshots', () => {
  test('Input component renders correctly', () => {
    const { container } = render(
      <TestWrapper>
        <Input placeholder='Test input' name='test' />
      </TestWrapper>
    )
    expect(container.firstChild).toMatchSnapshot('Input-default')
  })

  test('Input component with error state', () => {
    const { container } = render(
      <TestWrapper>
        <Input placeholder='Test input' errorMessage='This field is required' name='test-error' />
      </TestWrapper>
    )
    expect(container.firstChild).toMatchSnapshot('Input-error')
  })

  test('Button component renders correctly', () => {
    const { container } = render(
      <TestWrapper>
        <Button>Click me</Button>
      </TestWrapper>
    )
    expect(container.firstChild).toMatchSnapshot('Button-default')
  })

  test('Button component disabled state', () => {
    const { container } = render(
      <TestWrapper>
        <Button disabled>Disabled button</Button>
      </TestWrapper>
    )
    expect(container.firstChild).toMatchSnapshot('Button-disabled')
  })

  test('Pagination component renders correctly', () => {
    const { container } = render(
      <TestWrapper>
        <Pagination pageSize={20} queryConfig={{ page: '1' }} />
      </TestWrapper>
    )
    expect(container.firstChild).toMatchSnapshot('Pagination-default')
  })
})
