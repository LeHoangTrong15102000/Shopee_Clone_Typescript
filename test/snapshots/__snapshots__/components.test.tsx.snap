// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Snapshots > Button component disabled state > Button-disabled 1`] = `
<button
  class="cursor-not-allowed"
  disabled=""
>
  <span>
    Disabled button
  </span>
</button>
`;

exports[`Component Snapshots > Button component renders correctly > Button-default 1`] = `
<button
  class=""
>
  <span>
    Click me
  </span>
</button>
`;

exports[`Component Snapshots > Input component renders correctly > Input-default 1`] = `
<div>
  <input
    aria-invalid="false"
    class="w-full rounded-sm border border-gray-300 p-3 shadow-sm outline-none focus:border-gray-500"
    name="test"
    placeholder="Test input"
  />
  <div
    class="mt-1 min-h-[1.25rem] text-sm text-red-600"
  />
</div>
`;

exports[`Component Snapshots > Input component with error state > Input-error 1`] = `
<div>
  <input
    aria-describedby="test-error-error"
    aria-invalid="true"
    class="w-full rounded-sm border border-red-600 p-3 shadow-sm outline-none focus:border-gray-500"
    name="test-error"
    placeholder="Test input"
  />
  <div
    class="mt-1 min-h-[1.25rem] text-sm text-red-600"
    id="test-error-error"
  >
    This field is required
  </div>
</div>
`;

exports[`Component Snapshots > Pagination component renders correctly > Pagination-default 1`] = `
<nav
  aria-label="Pagination Navigation"
  class="mt-6 flex flex-wrap items-center justify-center"
  role="navigation"
>
  <span
    class="flex cursor-not-allowed items-center justify-center rounded-tl-sm rounded-bl-sm border-transparent px-4 py-2 "
  >
    <svg
      class="h-5 w-5 text-[rgba(0,0,0,0.6)]"
      fill="none"
      stroke="currentColor"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M15.75 19.5L8.25 12l7.5-7.5"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </span>
  <a
    class="flex cursor-pointer items-center justify-center px-4 py-2 text-[18px] bg-[#ee4d2d] text-white hover:bg-[#ee4d2d]"
    href="/?page=1"
  >
    1
  </a>
  <a
    class="flex cursor-pointer items-center justify-center px-4 py-2 text-[18px] border-transparent text-black/50 hover:text-[#ee4e2d]"
    href="/?page=2"
  >
    2
  </a>
  <a
    class="flex cursor-pointer items-center justify-center px-4 py-2 text-[18px] border-transparent text-black/50 hover:text-[#ee4e2d]"
    href="/?page=3"
  >
    3
  </a>
  <span
    class="flex items-center justify-center border bg-white px-3 py-2 shadow-sm"
  >
    ...
  </span>
  <a
    class="flex cursor-pointer items-center justify-center px-4 py-2 text-[18px] border-transparent text-black/50 hover:text-[#ee4e2d]"
    href="/?page=19"
  >
    19
  </a>
  <a
    class="flex cursor-pointer items-center justify-center px-4 py-2 text-[18px] border-transparent text-black/50 hover:text-[#ee4e2d]"
    href="/?page=20"
  >
    20
  </a>
  <a
    aria-label="Go to next page"
    class="flex cursor-pointer items-center justify-center rounded-tr-sm rounded-br-sm border-transparent  px-4 py-2 "
    href="/?page=2"
  >
    <svg
      class="h-5 w-5 text-[rgba(0,0,0,.6)] "
      fill="none"
      stroke="currentColor"
      stroke-width="1.5"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8.25 4.5l7.5 7.5-7.5 7.5"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </a>
</nav>
`;
